File: scripts/csv-to-json.js
const fs = require("fs");
const path = require('node:path');
const { parse } = require("csv-parse/sync");

const numColumns = 8;

// Time series
// 50 volts
// Ignore first 6 seconds.

const srcdir = "./recordings";
const destdir = "./app/data";

// https://docs.openbci.com/Cyton/CytonDataFormat/#interpreting-the-eeg-data
//const gain = 1;
//const scale = 4.5 / gain / (2^23 - 1);

const processFile = sourcefile => {
  const raw = fs.readFileSync(sourcefile);
  let rows = parse(
    raw,
    {
      columns: true,
      comment: "%",
      skip_empty_lines: true,
    }
  );

  rows = rows.map(row => ({
    timestamp: parseFloat(row[" Timestamp"].trim()),
    data: [
      parseFloat(row[' EXG Channel 0'].trim()),
      parseFloat(row[' EXG Channel 1'].trim()),
      parseFloat(row[' EXG Channel 2'].trim()),
      parseFloat(row[' EXG Channel 3'].trim()),
      parseFloat(row[' EXG Channel 4'].trim()),
      parseFloat(row[' EXG Channel 5'].trim()),
      parseFloat(row[' EXG Channel 6'].trim()),
      parseFloat(row[' EXG Channel 7'].trim()),
    ]}));

  const minValues = [9999999999, 9999999999, 9999999999, 9999999999, 9999999999, 9999999999, 9999999999, 9999999999];
  const maxValues = [0, 0, 0, 0, 0, 0, 0, 0];

  rows.forEach(row => {
    for (let i = 0; i < numColumns; i++) {
      const columnValue = row.data[i];
      if (columnValue < minValues[i]) {
        minValues[i] = columnValue;
      } else if (columnValue > maxValues[i]) {
        maxValues[i] = columnValue;
      }
    }
  });

  const scales = [];
  for (let i = 0; i < numColumns; i++) {
    scales[i] = 1.0 / (maxValues[i] - minValues[i]);
  }

  rows = rows.map(row => ({
    timestamp: row.timestamp,
    data: [
      (row.data[0] - minValues[0]) * scales[0],
      (row.data[1] - minValues[1]) * scales[1],
      (row.data[2] - minValues[2]) * scales[2],
      (row.data[3] - minValues[3]) * scales[3],
      (row.data[4] - minValues[4]) * scales[4],
      (row.data[5] - minValues[5]) * scales[5],
      (row.data[6] - minValues[6]) * scales[6],
      (row.data[7] - minValues[7]) * scales[7],
    ]}));

  return JSON.stringify(rows);
};

fs.readdirSync(srcdir).forEach(srcfilename => {
  const srcfilepath = path.join(srcdir, srcfilename);
  const data = processFile(srcfilepath);
  const destfilepath = path.join(
    destdir,
    `${path.basename(srcfilename, ".csv")}.json`
  );
  fs.writeFileSync(destfilepath, data);
});

File: scripts/csv-to-json.py
import os
import numpy as np
import pandas as pd

#import matplotlib
#matplotlib.use('Agg')

from brainflow.board_shim import BoardShim, BrainFlowInputParams, LogLevels, BoardIds
from brainflow.data_filter import DataFilter, FilterTypes, AggOperations, NoiseTypes

board_id = BoardIds.CYTON_BOARD.value

srcdir = "./recordings"
destdir = "./app/data"

def saveJSON(filename, data):
    destfilename = os.path.join(destdir, os.path.basename(filename).removesuffix(".csv") + ".json")
    with open(destfilename, 'w') as out:
        print("[", file=out)
        # SKIP FIRST HUNDRED ROWS until values settle
        print(",".join([f"{{\"timestamp\": {row[8]}, \"data\": [{', '.join(['%f' % n for n in row[:8]])}]}}" for row in data[100:]]), file=out)
        print("]", file=out)

def deexp(value):
    return pd.to_numeric(value.lower())

def processFile(filename):
    # skiprows includes comments.
    # We skip the comments, the column names, and the first sample, which is all zeros.
    # 21 is timestamp, 1..8 are eeg channels
    #data = np.array(np.loadtxt(filename, usecols=(21,1,2,3,4,5,6,7,8), delimiter=",", comments="%", skiprows=6))
    data = pd.read_csv(filename, delimiter=",", skiprows=6, usecols=[1,2,3,4,5,6,7,8,22], converters={22: deexp}).to_numpy()
    for channel in range(0, 8):
        #DataFilter.perform_bandpass(data[channel], 250, 2.0, 50.0, 4, FilterTypes.BESSEL.value, 0)
        #DataFilter.perform_highpass(data[channel], BoardShim.get_sampling_rate(board_id), 2.0, 4,
        #                            FilterTypes.BUTTERWORTH.value, 0)
        minVal = np.min(data[:,channel])
        maxVal = np.max(data[:,channel])
        rangeVal = maxVal - minVal
        if rangeVal > 0:
            scaleVal = 1.0 / rangeVal
            data[:,channel] = (data[:,channel] - minVal) * scaleVal
        else:
            data[:,channel] = 0
        #print(data[:,channel])
    if False:
        import matplotlib.pyplot as plt
        df = pd.DataFrame(data[100:])
        plt.figure()
        df.plot()
        plt.show()
    saveJSON(filename, data)

for srcfilename in os.listdir(srcdir):
    print(srcfilename)
    processFile(os.path.join(srcdir, srcfilename))
